src/main/scala/tech/sourced/gemini/FileQuery.scala

```diff
@@ -0,0 +1,153 @@
+import gopkg.in.bblfsh.sdk.v1.uast.generated.Node
+import org.bblfsh.client.BblfshClient
+import org.slf4j.{Logger => Slf4jLogger}
+import tech.sourced.featurext.FEClient
+import tech.sourced.featurext.generated.service.Feature
+import tech.sourced.featurext.generated.service.FeatureExtractorGrpc.FeatureExtractor
+
+import scala.collection.JavaConverters._
+import scala.util.control.NonFatal
+
+/**
+  * QueryResult contains iterators for full duplicates and similar files
+  *
+  * @param duplicates
+  * @param similar
+  */
+case class QueryResult(duplicates: Iterable[RepoFile], similar: Iterable[RepoFile])
+
+/**
+  * FileQuery queries similar and duplicated files
+  *
+  * @param conn db connection
+  * @param bblfshClient
+  * @param feClient
+  * @param docFreqPath will be replaced by database in future
+  * @param log
+  * @param keyspace
+  * @param tables
+  */
+class FileQuery(conn: Session,
+            bblfshClient: BblfshClient,
+            feClient: FeatureExtractor,
+            docFreqPath: String = "",
+            log: Slf4jLogger,
+            keyspace: String,
+            tables: Tables) {
+
+  /**
+    * find similar and duplicates for a file
+    *
+    * @param file
+    * @return
+    */
+  def find(file: File): QueryResult = {
+    val duplicates = findDuplicatesOfFile(file)
+    val duplicatedShas = duplicates.map(_.sha).toSeq
+
+    val similarShas = findSimilarForFile(file, conn, bblfshClient, feClient, docFreqPath)
+      .map(_.split("@")(1)) // value for sha1 in Apollo hashtables is 'path@sha1', but we need only hashes
+      .filterNot(sha => duplicatedShas.contains(sha))
+
+    log.info(s"${similarShas.length} SHA1's found to be similar")
+
+    val similar = similarShas.flatMap(sha1 => Database.repoFilesByHash(sha1, conn, keyspace, tables))
+    QueryResult(duplicates, similar)
+  }
+
+  protected def findDuplicatesOfFile(file: File): Iterable[RepoFile] = {
+    Database.repoFilesByHash(Gemini.computeSha1(file), conn, keyspace, tables)
+  }
+
+  protected def findSimilarForFile(
+                          file: File,
+                          conn: Session,
+                          bblfshClient: BblfshClient,
+                          feClient: FeatureExtractor,
+                          docFreqPath: String): Seq[String] = {
+    val docFreqFile = new File(docFreqPath)
+    if (!docFreqFile.exists()) {
+      log.warn("Document frequency for weighted min hash wasn't provided. Skip similarity query")
+      Seq()
+    } else {
+      extractUAST(file, bblfshClient) match {
+        case Some(node) =>
+          val featuresList = extractFeatures(feClient, node)
+          findSimilarFiles(featuresList, conn, docFreqFile)
+        case _ => Seq()
+      }
+    }
+  }
+
+  private def findSimilarFiles(featuresList: Iterable[Feature], conn: Session, docFreqFile: File): Seq[String] = {
+    if (featuresList.isEmpty) {
+      log.warn("file doesn't contain features")
```

Comment:
Just for the sake of ability to debug it from the logs, let's include a particular filename in this message.
Could be done either by passing a original filename in this function or by pushing the check to the client.

Reply:
Ok. But sorry what logs? It's query. You run it as cli and pass file there.